src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,161 @@
+    private static final String PLACEHOLDER_REGEX = "(.*)";
+
+    // identifies a single level wildcard character in the mqtt pattern. E.g. sensors/+/data
+    private static final String MQTT_TOPIC_SINGLE_LEVEL_WILDCARD_CHARACTER = "+";
+
+    // Regex expression used to replace the + in the mqtt pattern.
+    private static final String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+
+    // identifies a multi level wildcard character in the mqtt pattern. E.g. sensors/#
+    private static final String MQTT_TOPIC_MULTI_LEVEL_WILDCARD_CHARACTER = "#";
+
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static final String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+    private final List<MappingRule> rules;
+    private final List<Pattern> patterns = new ArrayList<>();
+
+    /**
+     * Constructor
+     * <p>
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(List<MappingRule> rules) {
+        this.rules = rules;
+        this.rules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
+        buildRegex();
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
+            HashMap<String, String> placeholders = new HashMap<>();
+
+            if (matcher.matches()) {
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+
+                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
+                Matcher placeholderMatcher = patterns.get(patterns.size() - 1).matcher(rule.getKafkaTopicTemplate());
+                while (placeholderMatcher.find()) {
+                    placeholders.put(placeholderMatcher.group(), null);
+                }
+
+                for (String placeholderKey : mqttTopicPatternParts) {
+                    if (placeholderKey.matches(MQTT_TOPIC_PLACEHOLDER_REGEX) && placeholders.containsKey(placeholderKey)) {
+                        placeholders.put(placeholderKey, matcher.group(removeBrackets(placeholderKey)));
+                    }
+                }
+
+                for (Map.Entry<String, String> entry : placeholders.entrySet()) {
+                    if (entry.getValue() != null) {
+                        mappedKafkaTopic = mappedKafkaTopic.replace(entry.getKey(), entry.getValue());
+                    } else {
+                        throw new IllegalArgumentException("The placeholder " + entry.getKey() + " was not assigned any value.");
+                    }
+                }
```

Comment:
Would it make sense to scan for missing placeholder values during the initial scan of the MappingRules when we are building our Pattern list and throw the exception then?

Reply:
@kyguy let me think about.